### PR TITLE
New lookup plugin : ini

### PR DESCRIPTION
**Pull request by Yannig**
Can handle ini file and java properties file. Can also read a list of value in a section using regexp.
Here's a simple example :

``` yaml
- name: "read properties value"
  set_fact:
    # java properties
    test1:    "{{lookup('ini', 'value1 type=properties  file=lookup.properties')}}"
    # ini file
    test2:    "{{lookup('ini', 'value1 section=global   file=lookup.ini')}}"
    # all value from section matching regexp value[1-2]
    value_section:    "{{lookup('ini', 'value[1-2] section=value_section file=lookup.ini re=true')}}"
- debug: var={{item}}
  with_items: [ 'test1', 'test2', 'value_section' ]
```

This pull request refer to #11019 (csvfile lookup plugin can now handle properties file).
Here's the result of the test playbook bundle in the pull request:

```
ansible-playbook test/integration/test_lookup_properties.yml

PLAY [Lookup test] ************************************************************ 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [read properties value] ************************************************* 
ok: [localhost]

TASK: [debug var={{item}}] **************************************************** 
ok: [localhost] => (item=test1) => {
    "item": "test1",                                                                                                                                                            
    "var": {                                                                                                                                                                    
        "test1": "Text associated with value1"                                                                                                                                  
    }                                                                                                                                                                           
}                                                                                                                                                                               
ok: [localhost] => (item=test2) => {
    "item": "test2",                                                                                                                                                            
    "var": {                                                                                                                                                                    
        "test2": "Same for value2"                                                                                                                                              
    }                                                                                                                                                                           
}                                                                                                                                                                               
ok: [localhost] => (item=test_dot) => {
    "item": "test_dot",                                                                                                                                                         
    "var": {                                                                                                                                                                    
        "test_dot": "Properties with dot"                                                                                                                                       
    }                                                                                                                                                                           
}                                                                                                                                                                               
ok: [localhost] => (item=field_with_space) => {
    "item": "field_with_space",                                                                                                                                                 
    "var": {                                                                                                                                                                    
        "field_with_space": "another space"                                                                                                                                     
    }                                                                                                                                                                           
}                                                                                                                                                                               

TASK: [read ini value] ******************************************************** 
ok: [localhost]

TASK: [debug var={{item}}] **************************************************** 
ok: [localhost] => (item=value1_global) => {
    "item": "value1_global",                                                                                                                                                    
    "var": {                                                                                                                                                                    
        "value1_global": "Text associated with value1 and global section"                                                                                                       
    }                                                                                                                                                                           
}                                                                                                                                                                               
ok: [localhost] => (item=value2_global) => {
    "item": "value2_global",                                                                                                                                                    
    "var": {                                                                                                                                                                    
        "value2_global": "Same for value2 and global section"                                                                                                                   
    }                                                                                                                                                                           
}                                                                                                                                                                               
ok: [localhost] => (item=value1_section1) => {
    "item": "value1_section1",                                                                                                                                                  
    "var": {                                                                                                                                                                    
        "value1_section1": "Another value for section1"                                                                                                                         
    }                                                                                                                                                                           
}                                                                                                                                                                               

TASK: [read ini value with section and regexp] ******************************** 
ok: [localhost]

TASK: [debug var={{item}}] **************************************************** 
ok: [localhost] => (item=value_section) => {
    "item": "value_section",                                                                                                                                                    
    "var": {                                                                                                                                                                    
        "value_section": "1,2"                                                                                                                                                  
    }                                                                                                                                                                           
}                                                                                                                                                                               
ok: [localhost] => (item=other_section) => {
    "item": "other_section",                                                                                                                                                    
    "var": {                                                                                                                                                                    
        "other_section": "4,5"                                                                                                                                                  
    }
}

TASK: [Reading unknown value] ************************************************* 
fatal: [localhost] => Failed to template {{lookup('ini', 'value2           section=section1 file=lookup.ini')}}: Unexpected error in during lookup: No option 'value2' in section: 'section1'

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/drayan/test_lookup_properties.retry

localhost                  : ok=7    changed=0    unreachable=1    failed=0   
```

### CODE REVIEW

**Comment by Yannig** [comment]
Forget to say that you can also use it in a with_ini context. Here's a simple example :

``` yaml
- name: "All value starting with 'value' in global section file ini"
  debug: var={{item}}
  with_ini: value.* file=lookup.ini section=global re=true

- name: "All value containing 'space' in global section file properties"
  debug: var={{item}}
  with_ini: .*space.* type=properties file=lookup.properties re=true
```

Here's the result:

``` yaml
TASK: [All value starting with 'value' in global section file ini] ************ 
ok: [localhost] => (item=Text associated with value1 and global section) => {
    "item": "Text associated with value1 and global section", 
    "var": {
        "Text": "Text"
    }
}
ok: [localhost] => (item=Same for value2 and global section) => {
    "item": "Same for value2 and global section", 
    "var": {
        "Same": "Same"
    }
}
ok: [localhost] => (item=Properties with dot) => {
    "item": "Properties with dot", 
    "var": {
        "Properties": "Properties"
    }
}

TASK: [All value containing 'space' in global section file properties] ******** 
ok: [localhost] => (item=another space) => {
    "item": "another space", 
    "var": {
        "another": "another"
    }
}
```

**Comment by Yannig** [comment]
Any comments regarding this pull request? Is there's something that need to be done?

**Comment by bcoca** [comment]
no, sorry, just need some time to be able to give a final review, also need to check if it is v2 compatible (there were some changes on how lookups work).

**Comment by Yannig** [comment]
Thanks for the feedback. I will give a try on a recent version of ansible v2.

**Comment by amenonsen** [comment]
@Yannig if you're rebasing to apply to v2 (i.e. current devel), please also add some documentation.

**Comment by Yannig** [comment]
@bcoca: rebase done. The plugin is working against last version of ansible devel.
@amenonsen: just finish to add documentation.
The first commit is just about a bad behaviour when the value does not exist: before this commit, ansible was exiting saying the value cannot be read from file. Now, we use the default value (just like in CSV plugin).

**Comment by Yannig** [comment]
@bcoca: rebase done. The plugin is working against last version of ansible devel.
@amenonsen: just finish to add documentation.
The first commit is just about a bad behaviour when the value does not exist: before this commit, ansible was exiting saying the value cannot be read from file. Now, we use the default value (just like in CSV plugin).

File: lib/ansible/plugins/lookup/ini.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from ansible import utils, errors
+import StringIO
+import os
+import codecs
+import ConfigParser
+import re
+
+class LookupModule(object):
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/ini.py]: this needs to be ported to version 2.0 type lookups, as is it won't load

**Comment by Yannig** [comment]
@bcoca: sorry, I miss a commit. As you can see, I'm very confortable with git...

**Pull request merged**